contracts/periphery/contracts/s2s/test/Transact.sol

```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.6.0;
+
+import "../SmartChainXLib.sol";
+import "../types/PalletEthereum.sol";
+
+// used for testing
+contract Transact {
+    function transact() public payable {
+        // 1. prepare the call that will be executed on the target chain
+        PalletEthereum.TransactCall memory call = PalletEthereum
+            .TransactCall(
+                // the call index of substrate_transact
+                0x2902,
+                // the evm transaction to transact
+                PalletEthereum.buildTransactionV2(
+                    0, // evm tx nonce, nonce on the target chain + pending nonce on the source chain + 1
+                    1000000000, // gasPrice, get from the target chain
+                    600000, // gasLimit, get from the target chain
```

Comment:
we can estimate `gasLimit`, but how can we decide the `nonce` and `gasPrice` on the target chain from the source chain?

Reply:
It need App SDK(darwinia.js)'s help.
- message.weight ≥ declare weight of the call, which is gas_to_weight(target_gas_limit)
- target_gas_price >= ThisRuntime::WeightToFee::calc(gas_to_weight(1)) * pallet_transaction_payment::Pallet::<ThisRuntime>::next_fee_multiplier() * 10**9
- target_evm_tx_nonce = nonce on the target chain + pending nonce on the source chain + 1

Reply:
We should pay attention the `value`, who pay the value?

Reply:
if `value` is set to non-zero, it means transfer from the sender address derived from dapp contract address. 
but if the  the sender address has no enough tokens, the `transact` will fail?

Reply:
what's the `next_fee_multiplier`? and will it be updated when the message come to the target chain?
On the target chain, the nonce and gasPrice belong to relayer. For the app layers, they have to know too much variable on the target chain. And it cannot work without `darwinia.js`.

Reply:
> if `value` is set to non-zero, it means transfer from the sender address on the target chain derived from dapp contract address on the source chain.
> but if the the sender address has no enough tokens, the `transact` will fail?

Yes, the tx would fail

Reply:
A use case is that, two users request at the same time. They get the same params from `darwinia.js`. Must one of them fail?

Reply:
yes, it needs `darwinia.js` or some tool  to work. In solidity, there is no way to get these variables automatically.

Reply:
> A use case is that, some users request at the same time. They get the same params from `darwinia.js`. Must most of them fail?

yes, one of them will fail.

Reply:
In the case, smart contract should store a `nonce` storage. And manage the `nonce` when do the cross call.

Reply:
I'm afraid that app developer will need some help from indexer to get the right nonce.

Just like sending ethereum tx will need help from explorer or pools or MetaMask.

Reply:
> In the case, smart contract should store a `nonce` storage. And manage the `nonce` when do the cross call.

Manage the `nonce` at source chain may be verify difficult, cause the cross call may fail.

Reply:
Message is delivered even if call failed, so it should not affect the nonce to increase.

Reply:
Ah, I see. That's a problem.

Reply:
I'll file another issue to track this, the solution would be it will always make sure nonce increased even it failed in validation and before final dispatch call.

Reply:
nonce shared by everyone?
or, one nonce sequence of one dapp contract

Reply:
> The nonce problem may cause frequent failures of cross-chain calls, because the nonce is not exclusive to the user, but shared by everyone.

Why? It is not shared.

Reply:
If nonce increased when it fails in validation, will affect the origin evm tx or not?

Reply:
> If message sender is a smart contract, there is a easy solution, contract should record/manage the last nonce in its contract storage.

Seems it also need the nonce increase when it fails.

Reply:
> If nonce increased when it fails in validation, will affect the origin evm tx or not?

It is the derived account on target chain, not the source account, there are only messages, no solo evm tx happened.

Reply:
> > The nonce problem may cause frequent failures of cross-chain calls, because the nonce is not exclusive to the user, but shared by everyone.
> 
> Why? It is not shared. You mean the same contract shared?

Yes, this remote nonce is a bit different from the ethereum local transaction nonce in that the initiator is not unique